crates/turbo-tasks/src/manager.rs

```diff
@@ -1487,23 +1488,24 @@ impl CurrentCellRef {
         tt.update_own_task_cell(
             self.current_task,
             self.index,
-            CellContent(Some(SharedReference::new(
-                Some(self.index.type_id),
-                triomphe::Arc::new(new_content),
-            ))),
+            CellContent(Some(SharedReference::new(triomphe::Arc::new(new_content)))),
         )
     }
 
-    pub fn update_shared_reference(&self, shared_ref: SharedReference) {
+    pub fn update_shared_reference(&self, shared_ref: TypedSharedReference) {
         let tt = turbo_tasks();
         let content = tt.read_own_task_cell(self.current_task, self.index).ok();
-        let update = if let Some(CellContent(Some(content))) = content {
-            content != shared_ref
+        let update = if let Some(TypedCellContent(_, CellContent(shared_ref_exp))) = content {
+            shared_ref_exp.as_ref().ne(&Some(&shared_ref.1))
         } else {
             true
         };
         if update {
-            tt.update_own_task_cell(self.current_task, self.index, CellContent(Some(shared_ref)))
+            tt.update_own_task_cell(
+                self.current_task,
+                self.index,
+                CellContent(Some(shared_ref.1)),
```

Comment:
destructing shared_ref at the beginning of the method would make it easier to read